Makefile

```
@@ -797,13 +795,27 @@ tools/doc/node_modules: tools/doc/package.json
 		cd tools/doc && $(call available-node,$(run-npm-ci)) \
 	fi
 
+RAWVER=$(shell $(PYTHON) tools/getnodeversion.py)
+VERSION=v$(RAWVER)
+
 .PHONY: doc-only
 doc-only: tools/doc/node_modules \
-	$(apidoc_dirs) $(apiassets) ## Build the docs with the local or the global Node.js binary.
+	$(apidoc_dirs)  ## Builds the docs with the local or the global Node.js binary.
 	@if [ "$(shell $(node_use_openssl_and_icu))" != "true" ]; then \
-		echo "Skipping doc-only (no crypto and/or no ICU)"; \
+		echo "Skipping doc-only (no crypto or no icu)"; \
 	else \
-		$(MAKE) out/doc/api/all.html out/doc/api/all.json out/doc/api/stability; \
+		$(call available-node, \
+			$(NPX) --prefix tools/doc api-docs-tooling generate \
+			-t legacy-html-all legacy-json-all api-links \
+			-i doc/api/*.md \
+			-i lib/*.js \
+			--ignore $(skip_apidoc_files) \
+			-o out/doc/api/ \
+			--no-lint \
+			-c file://$(PWD)/CHANGELOG.md \
+			-v $(VERSION) \
+			-p $(DOC_THREADS) \
+		) \
```

I wonder if it'd be better to let Make handle the caching and the parallelism – it would also decrease the maintenance burden of maintaining slightly different commands
```suggestion
		$(MAKE) $(apidocs_html) $(apidocs_json)
```

---

I'm generally -1 on this since the new tooling is designed to support handling all the files at once, but I'm either or atp. Wdyt @ovflowd?

---

> in general we want to generate all files…. But we should still support users that want to generate just one file or a set of files

The current setup allows for this, but the question was more for when we're generating all files, do we want to run the tooling separately for each file vs passing them all into the `-i` flag

---

I think my concern can be answered either by implementing the change I'm suggestion, or by explaining why it's not acceptable/desirable. In the end, it's not a deal-breaker, but I'd feel more comfortable approving this if I felt like I understood the tradeoffs.

---

I'm not sure I understand the concern, is the concern that starting the dev server takes time as it first builds everything?

---

The concern is that `make` is a system used to avoid rebuilding stuff that's already built, and I don't understand why we're not taking advantage of it – and yes, it's much much slower, I shared some figures in https://github.com/nodejs/node/pull/57343#pullrequestreview-2813551163.

---

> I'm not sure I understand the concern, is the concern that starting the dev server takes time as it first builds everything?

Yes, I believe so.

---

> The concern is that `make` is a system used to avoid rebuilding stuff that's already built, and I don't understand why we're not taking advantage of it – and yes, it's much much slower, I shared some figures in [#57343 (review)](https://github.com/nodejs/node/pull/57343#pullrequestreview-2813551163).

Noted. I'm totally fine if the point here is that Make helps skip unnecessary builds — but there's still a non-trivial cost in bootstrapping the tooling itself. Even just initializing everything takes some time. We also need context from other files to generate the docs — for instance, the navigation, search index, etc. need to be built using all the doc files.

And I’m not sure Make gives us enough fine-grained control at that level. Maybe what we actually want here is for the dev mode to behave more like ISR — and reduce the overhead during initialization.

One idea could be: parse all files upfront just enough to get the base AST, and then only run generators for files requested via the web server. WDYT @flakey5 @avivkeller — is that feasible? That way we’d be leveraging something like ISR to only generate what's actually needed, and we’d still watch for file changes.

Also wondering if we could cache some pieces of this — I know some build systems already do that, so reinventing the wheel might not be worth it. But I feel just having ISR-style logic might already help a lot, since the core parser runs relatively fast…

---

It's possible, see the linked issue in tooling (--select)

---

Maybe we do all the serving from tooling? That way, we can leverage whatever ISR-like capabilities we want?

But, I don't think that should block this, it's a QoL improvement, not something fundamental for distribution.

---

> there's still a non-trivial cost in bootstrapping the tooling itself

I feel like I'd be more than happy to pay this cost if that means I save 60s each time I run `make docserve` 😅 

> And I’m not sure Make gives us enough fine-grained control at that level

It does as long as you can store the intermediate info in intermediate files – but I'm not saying that necessarily the way to go.

---

I suggest we go with the `--select` method I outlined in the linked issue. This approach lets us keep the Makefile-style dependency management we use here—so we only rebuild the docs that have changed—while still supporting features that require access to all docs, like the sidebar.

It will involve some redundant parsing, since each doc will be processed individually with all the docs as context, but it should still be faster than the current 60s @aduh95 mentioned. Plus, we can add caching to store intermediate metadata and avoid unnecessary reprocessing.

---

Imo I say we just bite the bootstrapping costs and use make. If bootstrapping becomes a bottleneck then we can look into the options above like caching the AST output, etc. but I don't think it will. This probably will make the initial run of `doc-only` and rules that depend on it slower (due to bootstrapping) but afterwards it should be fine.

---

IMO the best compromise would be to have commands that are likely to be run by devs (e.g. `make docserve`) use `make` caching, and have commands used by CI skip it (e.g. `make test-ci`, `make test-doc-ci`); that way we would maximize the chances to hit the faster path.

---

Once https://github.com/nodejs/api-docs-tooling/pull/316/ lands, we can use the standard make parallelization without any problems